File: code/ch20/20.4.2.  289.gameOfLife.py
import copy


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        m = len(board)
        n = len(board[0])
        if m <= 0 or n <= 0:
            return []

        def cntLiveCell(i: int, j: int) -> int:
            cnt = 0
            directions = [
                (0, 1),
                (0, -1),
                (-1, 0),
                (1, 0),
                (1, 1),
                (1, -1),
                (-1, 1),
                (-1, -1),
            ]
            for (dx, dy) in directions:
                if i + dx >= 0 and i + dx < m and j + dy >= 0 and j + dy < n:
                    cnt += board[i + dx][j + dy] & 1
            return cnt

        for i in range(m):
            for j in range(n):
                # 八个方向有几个活细胞
                cnt = cntLiveCell(i, j)
                board[i][j] |= cnt << 1
        for i in range(m):
            for j in range(n):
                # 变化之前当前cell的值
                cell = board[i][j] & 1
                cnt = board[i][j] >> 1
                if cell == 0 and cnt == 3:
                    board[i][j] = 1
                elif cell == 1 and (cnt > 3 or cnt < 2):
                    board[i][j] = 0
                else:
                    board[i][j] = cell

File: code/ch20/20.4.2.  289.gameOfLife.java
public class Solution {
  public int cntLiveCell(int i, int j,
      int[][] board) {
    int cnt = 0;
    int[][] directions = {
      {0, 1},
      {0, -1},
      {-1, 0},
      {1, 0},
      {1, 1},
      {1, -1},
      {-1, 1},
      {-1, -1}
    };
    int m = board.length, n = board[0].length;
    for (int p = 0; p < 8; p++) {
      int dx = directions[p][0], dy = directions[p][1];
      if (i + dx >= 0 && i + dx < m && j + dy >= 0 &&
          j + dy < n)
        cnt += board[i + dx][j + dy] & 1;
    }
    return cnt;
  }
  public void gameOfLife(int[][] board) {
    int m = board.length;
    if (m <= 0) return;
    int n = board[0].length;
    if (n <= 0) return;
    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++) {
        int cnt = cntLiveCell(i, j, board);
        board[i][j] = board[i][j] | (cnt << 1);
      }
    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++) {
        int cell = board[i][j] & 1;
        int cnt = board[i][j] >> 1;
        if (cell == 0 && cnt == 3) board[i][j] = 1;
        else if (cell == 1 && (cnt > 3 || cnt < 2))
          board[i][j] = 0;
        else board[i][j] = cell;
      }
  }
}
